src/bind.py

```diff
@@ -227,25 +234,38 @@ def update_zonefiles_and_reload(
         logger.debug("Update and reload duration (ms): %s", (time.time_ns() - start_time) / 1e6)
 
     def _install_snap_package(
-        self, snap_name: str, snap_channel: str, refresh: bool = False
+        self, snap_name: str, snap_channel: str, snap_path: str, refresh: bool = False
     ) -> None:
         """Installs snap package.
 
         Args:
             snap_name: the snap package to install
             snap_channel: the snap package channel
+            snap_path: The path to the snap to install, can be blank.
             refresh: whether to refresh the snap if it's already present.
 
         Raises:
             InstallError: when encountering a SnapError or a SnapNotFoundError
         """
         try:
-            snap_cache = snap.SnapCache()
-            snap_package = snap_cache[snap_name]
-
-            if not snap_package.present or refresh:
-                snap_package.ensure(snap.SnapState.Latest, channel=snap_channel)
-
+            # If a snap resource was not given, install the snap as published on snapcraft
+            if snap_path == "":
+                snap_cache = snap.SnapCache()
+                snap_package = snap_cache[snap_name]
+
+                if not snap_package.present or refresh:
+                    snap_package.ensure(snap.SnapState.Latest, channel=snap_channel)
+            elif pathlib.Path(snap_path).is_file():
+                # Installing the charm via subprocess.
+                # Calling subprocess here is not a security issue.
+                subprocess.check_output(
+                    ["sudo", "snap", "install", snap_path, "--dangerous"]
+                )  # nosec
+            else:
+                logger.warning(
```

Comment:
minor/nitpick/not blocker: you could return at the beginning if snap_path has value but is not a file.
I feel this could be something like "if valid snap_path  install from file else install from store".

Reply:
After some discussions, I decided to simplify this function by relying on the exception that would be raised by the subprocess call.